### PR TITLE
[gitlabcomments] Add enricher to handle gitlab comments

### PR DESCRIPTION
**Pull request by vchrombie**
This PR proposes a new gitlab enricher to handle gitlab comments. The old enricher is left in place to avoid breaking changes with existing customer dashboards.

Ref: https://github.com/chaoss/grimoirelab/issues/208

### CODE REVIEW

**Comment by vchrombie** [comment]
The PR has a lot of pending work, I will be working on this. I will mark to ready for review after completing some good work.

Please let me know if you are having any suggestions.

**Comment by valeriocos** [comment]
Hi @vchrombie , please ping me when I can review this PR. I had a look at the current status and the start looks promising :)

WRT tests (if this can be of any help), you should run Perceval and copy some docs to a test file, which will be finally put in the folder [tests/data](https://github.com/chaoss/grimoirelab-elk/blob/master/tests/data/). The name is important since it is used to automatically identify the file and load its content (https://github.com/chaoss/grimoirelab-elk/blob/master/tests/base.py#L137)

If needed you can tweak the test data (e.g., changing value attributes) to make sure that all code is covered when running the tests.

**Comment by vchrombie** [comment]
Hi @valeriocos 
I have completed the major part of the PR. I am yet to complete the `__get_reactions` function and I am working on that right now.

**Comment by valeriocos** [comment]
Thanks for the update @vchrombie

File: grimoire_elk/enriched/gitlabcomments.py [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+        # extract reactions and add it to enriched item
+        # rich_mr.update(self.__get_reactions(merge_request))
+
+        rich_mr['merge_labels'] = merge_request['labels']
+        rich_mr['item_type'] = MERGE_TYPE
+
+        rich_mr['gitlab_repo'] = rich_mr['repository'].replace(GITLAB, '')
+        rich_mr['gitlab_repo'] = re.sub('.git$', '', rich_mr['gitlab_repo'])
+
+        # GMD code development metrics
```
Comment (vchrombie) [inline grimoire_elk/enriched/gitlabcomments.py:366]: I was building this enricher on top of `github2`.

GMD metrics implemented in _github2_ are
- number of forks (afaik, this is not possible in gitlab as we don't have the number of forks data)
- code merge duration (implemented :+1:)
- number of review comments (not done, but can be implemented by taking the number of `notes_data` of the MR WDYT?)

GMD metrics implemented in _gitlab_ are
- code merge duration (implemented :+1:)
- number of versions (done, will this be good to go here WDYT?)

I would like to know your comments on this. Let me know if I have to provide some perceval data to analyze.

File: grimoire_elk/enriched/gitlabcomments.py [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+            ecomment['issue_closed_at'] = eitem['issue_closed_at']
+            ecomment['gitlab_repo'] = eitem['gitlab_repo']
+            ecomment['repository'] = eitem['repository']
+            ecomment['item_type'] = COMMENT_TYPE
+            ecomment['sub_type'] = ISSUE_COMMENT_TYPE
+
+            # Copy data from the raw comment
+            ecomment['body'] = comment['body'][:self.KEYWORD_MAX_LENGTH]
+            ecomment['body_analyzed'] = comment['body']
+            # ecomment['url'] = comment['html_url']
```
Comment (vchrombie) [inline grimoire_elk/enriched/gitlabcomments.py:438]: I think we have to remove this as we don't have any comment URL in the data like in _github_.

File: grimoire_elk/enriched/gitlabcomments.py [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+            ecomment['merge_status'] = eitem['merge_status']
+            ecomment['gitlab_repo'] = eitem['gitlab_repo']
+            ecomment['repository'] = eitem['repository']
+            ecomment['item_type'] = COMMENT_TYPE
+            ecomment['sub_type'] = REVIEW_COMMENT_TYPE
+
+            # Copy data from the raw comment
+            ecomment['body'] = comment['body'][:self.KEYWORD_MAX_LENGTH]
+            ecomment['body_analyzed'] = comment['body']
+            # ecomment['url'] = comment['html_url']
```
Comment (vchrombie) [inline grimoire_elk/enriched/gitlabcomments.py:498]: same as above comment, think we need to remove this

File: grimoire_elk/enriched/gitlabcomments.py (hunk 1 above)
Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py:366]: > number of review comments (not done, but can be implemented by taking the number of notes_data of the MR WDYT?)

Perfect, agree on using notes_data

> number of versions (done, will this be good to go here WDYT?)

Yes, it makes sense to have it in the enriched index

File: grimoire_elk/enriched/gitlabcomments.py (hunk 2 above)
Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py:438]: Can we build a url with the information we have in the document? These URLs are good to allow navigating from the dashboard to the upstream item

File: grimoire_elk/enriched/gitlabcomments.py (hunk 3 above)
Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py:498]: Same comment as above

File: grimoire_elk/enriched/gitlabcomments.py [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+
+GEOLOCATION_INDEX = '/gitlab/'
+GITLAB = 'https://gitlab.com/'
+
+ISSUE_TYPE = 'issue'
+MERGE_TYPE = 'merge_request'
+
+COMMENT_TYPE = 'comment'
+ISSUE_COMMENT_TYPE = 'issue_comment'
+REVIEW_COMMENT_TYPE = 'review_comment'
```
Comment (vchrombie) [inline grimoire_elk/enriched/gitlabcomments.py:44]: Do you think we need to change this to 
```
MERGE_COMMENT_TYPE = 'merge_comment'
```
?

File: grimoire_elk/enriched/gitlabcomments.py (hunk 1 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:366]: Awesome, I will update the PR.

File: grimoire_elk/enriched/gitlabcomments.py (hunk 2 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:438]: Let me check and get back to you. :runner:

File: grimoire_elk/enriched/gitlabcomments.py (hunk 4 above)
Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py:44]: Yes, it is clear, thanks

File: grimoire_elk/enriched/gitlabcomments.py (hunk 2 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:438]: Hi @valeriocos 
I think we can reproduce the URL to these comments.

I am taking this MR as example, https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36.

In gitlab, every event is registered in the `notes_data` such as approving, commenting, merging, etc. I think we can consider all of them as they are provided with the API. I will address these events as comments itself.

So, each comment has an `id`. So we can build the URL using that `id` like `https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_id`.

I have provided some examples from the API data itself.

-  @yashk2000 approved this merge request : https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_325524652
-  Try formatting it properly as it is taking more characters in single line. Maybe from android studio you can use Ctl+Alt+L. : https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_325525231
-  @immadisairaj approved this merge request : https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_325529463

TL;DR 
It is feasible to construct the URL for comment using the `id`

Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py]: Can you check it once and let me know how it looks?

File: grimoire_elk/enriched/gitlabcomments.py (hunk 3 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py]: sam as above :D

File: grimoire_elk/enriched/gitlabcomments.py (hunk 1 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:366]: Done. :+1:

File: grimoire_elk/enriched/gitlabcomments.py (hunk 4 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py]: Done. :+1:

**Comment by vchrombie** [comment]
I will mark it ready-for-review once I complete the `__get_reactions` function too.

File: grimoire_elk/enriched/gitlabcomments.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+                continue
+
+            review_dates.append(str_to_datetime(comment['created_at']))
+
+        if review_dates:
+            return min(review_dates)
+
+        return None
+
+    def __get_reactions(self, item):
```
Comment (vchrombie) [inline grimoire_elk/enriched/gitlabcomments.py:197]: Hi @valeriocos 

I have completed implementing the `__get_reactions` function. 

gitlab reactions data is fetched a bit different than github. 
You can see the format here, [link](https://gist.github.com/vchrombie/a8b3449c924ccd01948d1a64d1d93be5).

The output of the current implementation looks something like this :point_down: 
```
{
    'reaction_thumbsup': 4, 
    'reaction_basketball': 2, 
    'reaction_thumbsdown': 1, 
    'reaction_rocket': 1, 
    'reaction_8ball': 1
}
```

Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py:197]: Since GitLab provides far many emojis than GitHub, I would suggest to add an attribute that counts the total number of reactions. To avod having too many mappings related to emojis (with the current implementation we will have a mapping for `reaction_thumbsup`, one for `reaction_8ball`, etc.), I would suggest to have just one mapping that contains the name/type of the emoji and the corresponding count (see example below). WDYT?
```
"reaction" : {
            "properties" : {
               "count" : {
                "type" : "long"
              },
               "type" : {
                "type" : "keyword"
              }
            }
          },
```

Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py:197]: current mappings:
```
          "reaction_100" : {
            "type" : "long"
          },
          "reaction_basketball_player" : {
            "type" : "long"
          },
          "reaction_bow" : {
            "type" : "long"
          },
          "reaction_bow_tone2" : {
            "type" : "long"
          },
          "reaction_bug" : {
            "type" : "long"
          },
          "reaction_call_me" : {
            "type" : "long"
          },
          "reaction_call_me_tone2" : {
            "type" : "long"
          },
          "reaction_cartwheel_tone3" : {
            "type" : "long"
          },
          "reaction_cartwheel_tone5" : {
            "type" : "long"
          },
          "reaction_clap" : {
            "type" : "long"
          },
          "reaction_clap_tone3" : {
            "type" : "long"
          },
          "reaction_confetti_ball" : {
            "type" : "long"
          },
          "reaction_cry" : {
            "type" : "long"
          },
          "reaction_eyes" : {
            "type" : "long"
          },
          "reaction_golfer" : {
            "type" : "long"
          },
          "reaction_grinning" : {
            "type" : "long"
          },
          "reaction_heart" : {
            "type" : "long"
          },
          "reaction_heart_eyes" : {
            "type" : "long"
          },
          "reaction_hearts" : {
            "type" : "long"
          },
          "reaction_heavy_plus_sign" : {
            "type" : "long"
          },
          "reaction_hugging" : {
            "type" : "long"
          },
          "reaction_joy" : {
            "type" : "long"
          },
          "reaction_kissing_smiling_eyes" : {
            "type" : "long"
....
```

File: grimoire_elk/enriched/gitlabcomments.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+        else:
+            rich_issue['time_open_days'] = rich_issue['time_to_close_days']
+
+        rich_issue['user_login'] = issue['author']['username']
+
+        user = issue.get('author', None)
+        if user is not None and user:
+            rich_issue['user_name'] = user['name']
+            rich_issue['author_name'] = user['name']
+            rich_issue["user_domain"] = self.get_email_domain(user['email']) if user['email'] else None
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:230]:
```
-            rich_issue["user_domain"] = self.get_email_domain(user['email']) if user['email'] else None
+            rich_issue["user_domain"] = self.get_email_domain(user['email']) if user.get('email', None) else None
```
Emails are currently not exposed by the GitLab API (AFAIK). We can keep the code, but we should make sure that `get_email_domain` is called only when `user['email']` exists.

File: grimoire_elk/enriched/gitlabcomments.py [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+            ecomment['gitlab_repo'] = eitem['gitlab_repo']
+            ecomment['repository'] = eitem['repository']
+            ecomment['item_type'] = COMMENT_TYPE
+            ecomment['sub_type'] = ISSUE_COMMENT_TYPE
+
+            # Copy data from the raw comment
+            ecomment['body'] = comment['body'][:self.KEYWORD_MAX_LENGTH]
+            ecomment['body_analyzed'] = comment['body']
+
+            ecomment['url'] = eitem['issue_url'] + '#note_' + comment['id']
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:452]:
```
-            ecomment['url'] = eitem['issue_url'] + '#note_' + comment['id']
+            ecomment['url'] = eitem['issue_url'] + '#note_' + str(comment['id'])
```

comment['id'] should be casted to str. We could use `format`, as done for calculating `ecomment['id']`

**Review by valeriocos — COMMENTED** [review]
Hi @vchrombie, thank you for advancing with this PR. I tested it on a battery of gitlab repos (taken from the gitlab:issue section at https://gitlab.com/Bitergia/c/gitlab/sources/-/blob/master/projects.json).

After fixing some minor errors (check comments), the execution was working fine.
I also checked that the url generated for comments work fine and allows to jump on the original comment on gitlab:
```
"sub_type" : "issue_comment",
          "body" : "@sabbap \"This merge request at least does something with the chevron until @JobV returns from his vacation when he can implement the scrolling.\"\r\n\r\n@JobV your call if this is good enough or not",
          "body_analyzed" : "@sabbap \"This merge request at least does something with the chevron until @JobV returns from his vacation when he can implement the scrolling.\"\r\n\r\n@JobV your call if this is good enough or not",
          "url" : "https://gitlab.com/gitlab-com/www-gitlab-com/issues/272#note_948621", <----
          "comment_updated_at" : "2015-03-13T00:16:20.314Z",
          "id" : "160818_merge_comment_948621",
          "grimoire_creation_date" : "2015-03-13T00:16:20.314000+00:00",
          "is_gitlab_issue_comment" : 1,
```

Can you add the tests and schema? Thanks

File: grimoire_elk/enriched/gitlabcomments.py (hunk 2 above)
Comment (valeriocos) [reply grimoire_elk/enriched/gitlabcomments.py]: Thank you, I have checked this while testing the connectors

File: grimoire_elk/enriched/gitlabcomments.py (hunk 7 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:452]: Perfect! :+1:

File: grimoire_elk/enriched/gitlabcomments.py (hunk 6 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:230]: Yes, this seems more better. :+1:

Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py]: I have updated the same for `assignee` and `merged_by` data too.

**Comment by vchrombie** [comment]
> https://github.com/chaoss/grimoirelab-elk/pull/881#pullrequestreview-425802955

Thanks for the review @valeriocos.
I have just updated the PR with the changes you have suggested.

> Can you add the tests and schema?

Sure, I will be working on it and add them here soon.

File: grimoire_elk/enriched/gitlabcomments.py (hunk 5 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:208]: I have corrected the implementation. Thanks for the help. :smile:

**Comment by vchrombie** [comment]
Recent updates to the PR
1. I removed the reactions mapping as we discussed it is not needed and also implemented the new `__get-reactions()` function.
2. added the schema using the gitlabcomments-mrs/issues_enriched index.
3. added the tests, (highly need a review on this as I'm not really good with it).

**Review by valeriocos — COMMENTED** [review]
Hi @vchrombie , thank you for working on this PR. Overall it looks good, I left some minor comments.

I also successfully executed the enricher on a short list of gitlab repos.

File: schema/gitlabcomments_issues.csv [hunk 8]
```diff
@@ -11,10 +11,10 @@ assignee_org_name,keyword,true,"Assignee organization name."
 assignee_user_name,keyword,true,"Assignee user name from SortingHat."
 assignee_uuid,keyword,true,"Assignee UUID from SortingHat."
 author_bot,boolean,true,"True if the given author is identified as a bot, from SortingHat profile."
-author_domain,keyword,true,"Domain associated to the author in SortingHat profile."
```
Comment (vchrombie) [inline schema/gitlabcomments_issues.csv]: Hi @valeriocos 
I was confused about this deletion.

I didn't see this (`author_domain`) in the mapping too, but `assignee_domain` and `merged_by_domain` are present.
Can you check test the enricher once and let me know if you have this?

The same deletion is present in the merge_request too.

**Comment by vchrombie** [comment]
Also, I have made a small script to generate schema file of an es index, [generate-es-index-schema-py](https://gist.github.com/vchrombie/bf6a682edcf47624126317897e58679c#file-generate-es-index-schema-py). 
This was built on top of @animeshk08's script as it had some backlogs. 

The script can be improved by adding the idea of default field and adding descriptions. Please leave your comments in the gist or you can open an issue here, [vchrombie/grimoirelab-scripts](https://github.com/vchrombie/grimoirelab-scripts). Feedback is highly welcomed. ^^

File: schema/gitlabcomments_issues.csv (hunk 8 above)
Comment (valeriocos) [reply schema/gitlabcomments_issues.csv]: `*_domain` fields are added by SortingHat. By default GitLab users don't have their emails exposed on the API. However, if you execute sortinghat with a matching algorithm that joins the identities based on username and email (the algorithms are available [here](https://github.com/chaoss/grimoirelab-sortinghat/tree/master/sortinghat/matching)), the email information can be collected. The `*_domain` fields should be filled and thus they will appear in the mapping.

Comment (vchrombie) [reply schema/gitlabcomments_issues.csv]: Okay, got it. So, I will add them back to the schema then.

Comment (valeriocos) [reply schema/gitlabcomments_issues.csv]: Perfect, let me know when I can review the PR, thanks!

**Comment by vchrombie** [comment]
Hi @valeriocos 
I have updated the PR as per your suggestions.
Can you review this when you are free?
Thanks.

File: grimoire_elk/enriched/gitlabcomments.py [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+        other than the user who created the issue
+        """
+        comments_dates = [str_to_datetime(comment['created_at']) for comment in item['notes_data']
+                          if item['author']['username'] != comment['author']['username']]
+        reaction_dates = [str_to_datetime(reaction['created_at']) for reaction in item['award_emoji_data']
+                          if item['author']['username'] != reaction['user']['username']]
+        reaction_dates.extend(comments_dates)
+
+        if reaction_dates:
+            return min(reaction_dates)
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:173]: To increase the test coverage use a single return. You can define a variable (e.g., `first_attention`), which is set to None and modified in this if statement

File: grimoire_elk/enriched/gitlabcomments.py [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+            # skip comments of the merge request creator
+            if item['author']['username'] == comment['author']['username']:
+                continue
+
+            review_dates.append(str_to_datetime(comment['created_at']))
+
+        if review_dates:
+            return min(review_dates)
+
+        return None
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:193]: Similar comment as above, please use just a return

File: grimoire_elk/enriched/gitlabcomments.py [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+        else:
+            rich_issue['time_open_days'] = rich_issue['time_to_close_days']
+
+        author = issue.get('author', None)
+        if author is not None and author:
+            rich_issue['author_login'] = author['username']
+            rich_issue['author_name'] = author['name']
+            rich_issue['author_domain'] = self.get_email_domain(author['email']) if author.get('email') else None
+        else:
+            rich_issue['author_login'] = None
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:228]: This else block isn't covered by the test data

File: grimoire_elk/enriched/gitlabcomments.py [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+        rich_issue['item_type'] = ISSUE_TYPE
+
+        rich_issue['gitlab_repo'] = rich_issue['repository'].replace(GITLAB, '')
+        rich_issue['gitlab_repo'] = re.sub('.git$', '', rich_issue['gitlab_repo'])
+
+        if self.prjs_map:
+            rich_issue.update(self.get_item_project(rich_issue))
+
+        if 'project' in item:
+            rich_issue['project'] = item['project']
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:268]: Not covered by the tests

File: grimoire_elk/enriched/gitlabcomments.py [hunk 13]
```diff
@@ -0,0 +1,573 @@
...
+        else:
+            rich_mr['time_open_days'] = rich_mr['time_to_close_days']
+
+        author = merge_request.get('author', None)
+        if author is not None and author:
+            rich_mr['author_login'] = author['username']
+            rich_mr['author_name'] = author['name']
+            rich_mr['author_domain'] = self.get_email_domain(author['email']) if author.get('email') else None
+        else:
+            rich_mr['author_login'] = None
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:310]: Not covered by the test data

File: grimoire_elk/enriched/gitlabcomments.py [hunk 14]
```diff
@@ -0,0 +1,573 @@
...
+        if merge_request['notes_data'] != 0:
+            min_review_date = self.get_time_to_merge_request_response(merge_request)
+            rich_mr['time_to_merge_request_response'] = \
+                get_time_diff_days(str_to_datetime(merge_request['created_at']), min_review_date)
+
+        if self.prjs_map:
+            rich_mr.update(self.get_item_project(rich_mr))
+
+        if 'project' in item:
+            rich_mr['project'] = item['project']
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:365]: Not covered by the tests

File: grimoire_elk/enriched/gitlabcomments.py [hunk 15]
```diff
@@ -0,0 +1,573 @@
...
+    @metadata
+    def get_rich_item(self, item):
+        rich_item = {}
+
+        if item['category'] == 'issue':
+            rich_item = self.__get_rich_issue(item)
+        elif item['category'] == 'merge_request':
+            rich_item = self.__get_rich_merge(item)
+        else:
+            logger.error("[gitlab] rich item not defined for GitLab category {}".format(
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:388]: Not covered by the tests

File: grimoire_elk/enriched/gitlabcomments.py [hunk 16]
```diff
@@ -0,0 +1,573 @@
...
+            num_items += len(items_to_enrich)
+            ins_items += self.elastic.bulk_upload(items_to_enrich, self.get_field_unique_id())
+            items_to_enrich = []
+
+        if len(items_to_enrich) > 0:
+            num_items += len(items_to_enrich)
+            ins_items += self.elastic.bulk_upload(items_to_enrich, self.get_field_unique_id())
+
+        if num_items != ins_items:
+            missing = num_items - ins_items
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:568]: Not covered by the test data

File: grimoire_elk/enriched/gitlabcomments.py [hunk 17]
```diff
@@ -0,0 +1,573 @@
...
+                eitems = self.enrich_issue(item, eitem)
+            elif item['category'] == MERGE_TYPE:
+                eitems = self.enrich_merge(item, eitem)
+
+            items_to_enrich.extend(eitems)
+
+            if len(items_to_enrich) < MAX_SIZE_BULK_ENRICHED_ITEMS:
+                continue
+
+            num_items += len(items_to_enrich)
```
Comment (valeriocos) [inline grimoire_elk/enriched/gitlabcomments.py:559]: Not covered by the tests

**Review by valeriocos — COMMENTED** [review]
Hi @vchrombie ,

The PR is almost ready to be merged, thank you for your work! Please check the comments (for the ones concerning the test coverage, try to address them if possible).

Please check the actions below too:
- create a yaml file to provide a high-level description of this new addition (https://github.com/Bitergia/release-tools#changelog). As a reference you can check the PR for rocketchat
- create an issue in https://github.com/Bitergia/release-tools#changelog to give feedback on the use of the changelog (you can check the issues 14 and 15 in that repo as a reference) 
- update the readme at https://github.com/chaoss/grimoirelab-elk#enriched-data
- submit a PR to mordred to document how to execute the gitlabcomments: https://github.com/chaoss/grimoirelab-sirmordred#supported-data-sources-
- squash the commits of this PR in just one
- sync this branch with master

File: grimoire_elk/enriched/gitlabcomments.py (hunk 9 above)
Comment (vchrombie) [reply grimoire_elk/enriched/gitlabcomments.py:173]: Nice hack.

**Comment by valeriocos** [comment]
please @vchrombie ping me when the PR is ready for review, thanks

**Comment by almereyda** [comment]
Coming from outside and having dismissed GrimoireLab previously for lacking GitLab support, I would still like to see it support additional code forges other than GitHub, as per the arguments in inhttps://github.com/chaoss/grimoirelab/issues/284#issuecomment-827155190).

Is there something me as an outsider can contribute to the formal aspects of the PR, in so the remaining technical questions can be answered more easily? Speaking of:

- [ ] rebase on master
- [ ] update elk readme
- [ ] add mordred supported data sources documentation (PR)
- [ ] update tests for edge cases and error conditions

**Comment by vchrombie** [comment]
Hi @almereyda, thanks for your comment.

> Coming from outside and having dismissed GrimoireLab previously for lacking GitLab support

Sorry about it, but the current GrimoireLab supports analyzing issues and merge requests from the gitlab projects ([reference](https://github.com/chaoss/grimoirelab-sirmordred#gitlab-)). This PR is an additional enricher which can analyze the comments of the gitlab issues and mrs. I couldn't complete it because I completely forgot about this.

> Is there something me as an outsider can contribute to the formal aspects of the PR, in so the remaining technical questions can be answered more easily? Speaking of:

Thanks for offering help, I appreciate it. But, I think I can complete the work on this PR. I can pull some time during the coming weekend and complete te pending work.
